cannon/mipsevm/tests/difftester.go

```diff
@@ -277,23 +296,63 @@ func (e normalExecResult) assertExpectedResult(t testing.TB, goVm mipsevm.FPVM,
 }
 
 type vmPanicResult struct {
-	panicMsg string
-	evmError string
+	panicValue           interface{}
+	evmErrorMatcher      testutil.ErrMatcher
+	memoryProofAddresses []arch.Word
+	proofData            []byte
+}
+
+type VMPanicTestOption func(*vmPanicResult)
+
+func WithProofData(proofData []byte) VMPanicTestOption {
+	return func(vmPanicResult *vmPanicResult) {
+		vmPanicResult.proofData = proofData
+	}
 }
 
-func ExpectVmPanic(goPanicMsg, evmRevertMsg string) ExpectedExecResult {
-	return vmPanicResult{
-		panicMsg: goPanicMsg,
-		evmError: evmRevertMsg,
+func WithMemoryProofAddr(addr arch.Word) VMPanicTestOption {
+	return func(vmPanicResult *vmPanicResult) {
+		vmPanicResult.memoryProofAddresses = append(vmPanicResult.memoryProofAddresses, addr)
 	}
 }
 
-func (e vmPanicResult) assertExpectedResult(t testing.TB, goVm mipsevm.FPVM, vmVersion VersionedVMTestCase, expect *mtutil.ExpectedState) {
+func ExpectVmPanic(goPanicValue interface{}, evmRevertMsg string, options ...VMPanicTestOption) ExpectedExecResult {
+	result := vmPanicResult{
+		panicValue:      goPanicValue,
+		evmErrorMatcher: testutil.StringErrorMatcher(evmRevertMsg),
+	}
+	for _, opt := range options {
+		opt(&result)
+	}
+	return result
+}
+
+func ExpectVmPanicWithCustomErr(goPanicMsg interface{}, customErrSignature string, options ...VMPanicTestOption) ExpectedExecResult {
+	result := vmPanicResult{
+		panicValue:      goPanicMsg,
+		evmErrorMatcher: testutil.CustomErrorMatcher(customErrSignature),
+	}
+	for _, opt := range options {
+		opt(&result)
+	}
+	return result
+}
+
+func (e vmPanicResult) assertExpectedResult(t testing.TB, goVm mipsevm.FPVM, vmVersion VersionedVMTestCase, expect *mtutil.ExpectedState, cfg *TestConfig) {
 	state := goVm.GetState()
-	proofData := vmVersion.ProofGenerator(t, state)
-	errMsg := testutil.CreateErrorStringMatcher(e.evmError)
-	testutil.AssertEVMReverts(t, state, vmVersion.Contracts, nil, proofData, errMsg)
-	require.PanicsWithValue(t, e.panicMsg, func() { _, _ = goVm.Step(false) })
+	proofData := e.proofData
+	if proofData == nil {
+		proofData = vmVersion.ProofGenerator(t, state, e.memoryProofAddresses...)
+	}
+	testutil.AssertEVMReverts(t, state, vmVersion.Contracts, cfg.tracingHooks, proofData, e.evmErrorMatcher)
+
+	if panicErr, ok := e.panicValue.(error); ok {
+		require.PanicsWithError(t, panicErr.Error(), func() { _, _ = goVm.Step(false) })
+	} else if panicStr, ok := e.panicValue.(string); ok {
+		require.PanicsWithValue(t, e.panicValue, func() { _, _ = goVm.Step(false) })
+	} else {
+		t.Fatalf("Invalid panic value provided.  Go panic value must be a string or error.  Got: %v", panicStr)
```

Comment:
```suggestion
		t.Fatalf("Invalid panic value provided.  Go panic value must be a string or error.  Got: %v", panicValue)
```
panicStr here is empty since it's not a string

Reply:
Good catch!